### ISSUE TITLE
around-next tap-macro-release: not releasing correctly

### ISSUE THREAD

**Issue by entibo** [kmonad/kmonad#576]
Suppose a button bound to `x`:

```clojure
(around-next (tap-macro-release u v))
```
Expected: `Tx Tk` produces `ukv`
Actual: `Tx Tk` produces `uk`; only when Kmonad receives **another** `Rx` does it emits `v`

---

#### For reference:
```clojure
(around (tap-macro-release u v) k)
```
`Tx` produces `ukv`

#### Use-case:
The goal was to wrap the *next* button with a *before* button and a *after* button

```clojure
(around-next (tap-macro-release C-f #(ret esc)))
```
`Tx Tk`: move cursor/focus to the next `k` character

#### Workaround:
```clojure
(around-next (momentary-layer verbose))
```
```clojure
(deflayer verbose
    ...
    (tap-macro u k v)
    ...
)
```
If `k` can be any key, this would of course be very impractical.

**Comment by david-janssen** [comment]
So I would like to respond to 2 things.

1. `around` works but `around-next` doesn't.

That's a good catch: ideally these two formulations should be the same, because they are indeed semantically the same. The fact that they are different must have something to do with `around-next` doing some timer-magic while trying to perform its task. I've come to the conclusion that timer-magic is probably black magic, and I'd like to redesign the core engine to get away from timers. The race conditions and interactions are a huge headache to think about and debug. This fix is OnThePile(tm), and near but not on the top.

2. you're not entirely correct about `around` and `around-next`

You describe trying to wrap some button in a before and a after button. The `around` family doesn't actually try to do that. It tries to wrap 1 button with 1 other button. I.e.
1. Press the wrapper
2. Tap the wrapped
3. Release the wrapper

Although actually, looking at the `around` + `tap-macro-release` (of 2 buttons) would achieve that effect, because it would trigger the 2nd button only after the wrapped got through. So maybe you're not 'not entirely correct', maybe you're just exceedingly clever. 

The use-case of wrapping 1 button in 2 separate buttons never really occured to me, but it should be relatively simple to support natively. Maybe I can add support for this. Lemme fuss about a bit.

**Comment by david-janssen** [comment]
Yeah, that was about as straightforward as I thought it was. Try: `(before-after-next a b)`. This button will:
1. tap `a` the moment you press it
2. wait for the next press of some other key
3. tap `b` immediately after release of *that key*

**Comment by entibo** [comment]
> 1. tap `a` the moment you press it  

This isn't necessarily desirable, and it can already be achieved with `#(a (around-next ...))`, so maybe it should be tapped when the next key is pressed (but before it) ?

Also, `before-after-next` is missing the `around` and `timeout` functionalities.  
There could be variants such as  `after-next-timeout`, `before-around-after-next`... but i have to admit i have no idea how those would be useful.

I feel like there could be a more powerful abstraction for dealing with *next* keys.  
Something like `next-key` and `next-key-timeout`:
```clojure
(around-next x)
(next-key (around x _))

(before-after-next a b)
#(a (next-key (around (tap-macro-release b) _)))

;; same but `a` is tapped only when the next key is pressed
(next-key (around (tap-macro-release a b) _))

;; same as above but b is held while the next key is pressed, instead of tapped at the end
(next-key #(a (around _ b)))

;; same as above but *only* `b` is held, not _
(next-key #(a _ b))

```

**Comment by david-janssen** [comment]
> I feel like there could be a more powerful abstraction for dealing with next keys

Absolutely, the current representation of keys is very poor. There are at least 2 different dimensions that can be modelled separately and then combined into their product. I.e.: there are patterns-we-match and actions-we-perform, and we currently code all of those as buttons. A much more expressive language is possible, and is very much my next big project after I get the IO-shell working very solidly.

In regards to the tweaks to the `before-after-next` button, maybe I can find some time soonish, but at the moment I'd rather focus on the big project :).

**Comment by jokesper** [comment]
#872 fixed some of those buttons, but generally KMonad is not equiped to handle buttons which use hooks inside other buttons.
